### ISSUE TITLE
Model Classification Bias: 95%+ frames labeled as "Labral Mobilization"

### ISSUE THREAD

**Issue by ncightadmin** [ncightadmin/video-analyzer#1]
## Problem Summary
The AI model successfully detects phase transitions (108 phases found) but incorrectly classifies 95%+ of them as "Labral Mobilization" instead of using all 8 phase types.

---

## Environment
- **Hardware:** MacBook Pro M4 Pro
- **OS:** macOS
- **Python Version:** 3.11
- **PyTorch Version:** 2.9.1
- **Repository:** Fork from original repo
- **Model File:** `model-inference/best_model.pt` (present and loads successfully)

---

### Test 1: First Surgical Video
- **Duration:** 17.9 minutes (1073 seconds)
- **Result:** All 108 detected phases labeled as "Labral Mobilization"
- **Events:** All 108 bleeding events labeled identically

### Test 2: Second Surgical Video  
- **Duration:** Similar length
- **Result:** 
  - ✅ Phase 1: "Diagnostic" (actually "Diagnostic Arthroscopy") - **CORRECT**
  - ❌ Phases 2-46: All "Labral Mobilization" - **INCORRECT**

---

### Terminal Log Excerpt
```
INFO:MasterInferenceEngine:Phase detector loaded successfully
INFO:MasterInferenceEngine:Instrument tracker loaded successfully
INFO:MasterInferenceEngine:Event detector loaded successfully
INFO:MasterInferenceEngine:Motion analyzer loaded successfully

INFO:MasterInferenceEngine:Integration completed: 108 phases, 72 instruments, 108 bleeding events
INFO:MasterInferenceEngine:Analysis complete in 88.06s
```

**Key Observations:**
- ✅ All model components load successfully
- ✅ Model processes video and detects 108 phase transitions
- ✅ Detection works - classification is broken

### Raw JSON Output
```json
{
    "phase_type": "SurgicalPhaseType.LABRAL_MOBILIZATION",
    "start_frame": 0,
    "end_frame": 60,
    "confidence": 0.9992648959159851
},
{
    "phase_type": "SurgicalPhaseType.LABRAL_MOBILIZATION",
    "start_frame": 300,
    "end_frame": 360,
    "confidence": 0.9978138208389282
}
```
Pattern repeats for all 108 phases.

### Excel Export Results

**Add screenshots here** (drag and drop your Excel screenshots showing the Phases tab)

---

### Code Investigation

**File:** `surgical_ai_system/models/phase_detection/phase_detector.py`

**Lines 58-61 - Phase Labels Definition:**
```python
phase_labels = [
    "Portal Placement", "Diagnostic Arthroscopy", "Labral Mobilization",
    "Glenoid Preparation", "Anchor Placement", "Suture Passage", 
    "Suture Tensioning", "Final Inspection"
]
```
✅ **Phase labels are correctly defined in code**

**Lines 100-106 - Prediction Logic:**
```python
predicted_class = torch.argmax(probabilities, dim=1).item()
confidence = probabilities[0][predicted_class].item()
predicted_phase = phase_labels[predicted_class]
```
✅ **Prediction logic is correct**

### The Problem

**The model consistently predicts class index 2** (which maps to "Labral Mobilization") for 95%+ of frames.

**Phase Label Array Mapping:**
- Index 0 → "Portal Placement"
- Index 1 → "Diagnostic Arthroscopy"  
- **Index 2 → "Labral Mobilization"** ← Model defaults to this
- Index 3 → "Glenoid Preparation"
- Index 4 → "Anchor Placement"
- Index 5 → "Suture Passage"
- Index 6 → "Suture Tensioning"
- Index 7 → "Final Inspection"

### Why This Happens

**Hypothesis: Class Imbalance During Training**

The neural network likely experienced:
1. **Unbalanced training data** - far more "Labral Mobilization" examples than other phases
2. **Overfitting to dominant class** - model learned to default to class 2 for best overall accuracy
3. **Insufficient examples** of other phases during training

**Evidence supporting this:**
- Model CAN predict other classes (correctly identified "Diagnostic Arthroscopy" in Test 2)
- High confidence scores (0.95-0.99) even for incorrect predictions
- Consistent bias across multiple different videos

---

### Current State
- ❌ Unable to accurately analyze surgical videos
- ❌ Phase detection unreliable for clinical/training use
- ❌ Event detection also shows monotony (all "bleeding"/"Moderate")
- ✅ Model architecture and code are functional
- ✅ Detection of phase boundaries works

### Use Case Impact
- Cannot provide accurate phase timelines
- Cannot generate reliable surgical reports
- Limited value for surgical training/review
- Demonstrations show poor model performance

---

### Solution 1: Updated Model Weights (Easiest - Immediate)
**Replace `best_model.pt` with better-trained model**

**Questions:**
- Is there an updated model file with better class balance?
- Was the model trained on balanced dataset or with class weights?
- Are there metrics showing per-class accuracy from training?

**Action:** Provide updated model file if available

---

### Solution 2: Confidence Thresholding (Medium - Temporary Fix)

**Modify prediction logic to reject low-confidence dominant class predictions.**

**Implementation:** (around line 103 in `phase_detector.py`)
```python
predicted_class = torch.argmax(probabilities, dim=1).item()
confidence = probabilities[0][predicted_class].item()

# If predicting Labral Mobilization with marginal confidence, use second choice
if predicted_class == 2 and confidence < 0.85:
    probabilities[0][2] = 0  # Mask dominant class
    predicted_class = torch.argmax(probabilities[0]).item()
    confidence = probabilities[0][predicted_class].item()

predicted_phase = phase_labels[predicted_class]
```

**Pros:** Quick fix, no retraining needed  
**Cons:** Workaround, doesn't address root cause  

---

### Solution 3: Model Retraining (Long-term - Proper Fix)

**Retrain with balanced dataset and proper techniques:**

1. **Collect balanced training data** 
   - Equal representation of all 8 phases
   - Multiple surgeons, conditions, camera angles

2. **Apply class weights during training**
```python
   class_weights = compute_class_weight('balanced', 
                                        classes=np.unique(labels), 
                                        y=labels)
   criterion = nn.CrossEntropyLoss(weight=torch.FloatTensor(class_weights))
```

3. **Data augmentation** for under-represented classes

4. **Monitor per-class metrics** during training
   - Per-class accuracy
   - Confusion matrix
   - F1 scores for each phase

**Pros:** Proper solution, better overall accuracy  
**Cons:** Time-intensive, requires more training data  

---

## Questions for Development Team

1. **Model Training:**
   - What was the phase distribution in the training dataset?
   - Were class weights or sampling techniques used to handle imbalance?
   - What were the per-class accuracy metrics during training/validation?

2. **Model Availability:**
   - Is there an updated `best_model.pt` file available?
   - Are there different model versions we can test?
   - What's the expected baseline accuracy for each phase type?

3. **Configuration:**
   - Are there any configuration files that control classification behavior?
   - Should MPS (GPU) acceleration be enabled for M4 Pro? (Currently unclear if active)
   - Are there hyperparameters we can adjust without retraining?

4. **Validation:**
   - What testing videos/datasets do you use for validation?
   - What's the expected accuracy on those datasets?
   - Can you share validation results showing per-class performance?

---

### Model Loading (Successful)
```
Phase detector loaded successfully
Instrument tracker loaded successfully  
Event detector loaded successfully
Motion analyzer loaded successfully
```

### Performance Metrics
- Analysis time: 88 seconds for 17.9-minute video
- Frame sampling: 110 samples from 32,200 frames (every 300 frames)
- Confidence scores: 0.95-0.99 (very high, indicating model certainty)

### System Setup
- Virtual environment: `venv_desktop` with Python 3.11
- All dependencies installed from `requirements.txt`
- Model file present at: `model-inference/best_model.pt`
- GPU acceleration: MPS available (M4 Pro) but unclear if active

---

## Next Steps

**Immediate Actions Needed:**
1. Confirm if updated model file is available
2. Provide training dataset statistics (phase distribution)
3. Share validation metrics showing per-class accuracy
4. Advise on whether to implement confidence thresholding as temporary fix

**For Testing:**
- Can provide additional test videos if needed
- Can test different model files or configurations
- Can validate any fixes/updates

---

## Additional Context

**Setup Process:**
- Forked repository successfully
- Cloned to local Mac M4 Pro
- Installed Python 3.11 (3.13 had tkinter issues)
- Installed python-tk@3.11 for GUI support
- All requirements installed without errors
- Desktop GUI launches and runs successfully

**Files Generated:**
- JSON analysis results
- Excel comprehensive reports
- CSV metrics
- Phase timeline visualizations
- All files saved correctly

The infrastructure works perfectly - we just need a better-trained model or guidance on improving classification accuracy.
```

---

## 📸 Adding Your Screenshots

**After pasting the issue text:**

1. **Find this section:** "### Excel Export Results"
2. **Below "Add screenshots here"**
3. **Drag and drop** your Excel screenshots showing:
   - Phases tab with all "Labral Mobilization"
   - Events tab showing all "bleeding"
4. GitHub will automatically upload and embed them

---

## ✅ Before Submitting - Quick Checklist

- [ ] Title is clear and descriptive
- [ ] Description is pasted in
- [ ] Screenshots are attached (optional but recommended)
- [ ] You've reviewed for any sensitive information
- [ ] Ready to submit!

---

## 📤 After Submitting

1. **GitHub will assign an issue number** (e.g., #1, #2, etc.)
2. **Copy the issue URL** (looks like: `https://github.com/ncightadmin/video-analyzer/issues/1`)
3. **Send to your dev team** via:
   - Email with the link
   - Slack/Teams message
   - However you normally communicate

---

## 💡 Pro Tips

**If you want to:**
- **Assign to someone:** Use "Assignees" on the right side
- **Add labels:** Use "Labels" (like "bug", "priority", etc.)
- **Tag team members:** Use `@username` in the description

**Example:**
```
@devteam-lead This is blocking our surgical video analysis workflow. Please prioritize.

**Comment by muhammadanas25** [comment]
Can you specify on which video file this issue was reported so I could reproduce it and that would help me better to understand and solve this issue?
Also can you use the web app to do inference ? its is the same web app  which I shown in demo
the desktop app is currently using the older model thats why you are getting wrong labels let me fix that meanwhile you can try inference using web app

**Comment by ncightadmin** [comment]
I shared the video that I analyzed into the shared video folder on dropbox under the "Issues Folder"Alonzo Sexton, MDCEO, Founder
        
        
          On Dec 18 2025, at 8:13 pm, muhammadanas25 ***@***.***> wrote:
          
        
        
           muhammadanas25 left a comment (ncightadmin/video-analyzer#1)
Can you specify on which video file this issue was reported so I could reproduce it and that would help me better to understand and solve this issue?
Also can you use the web app to do inference ? its is the same web app  which I shown in demo
the desktop app is currently using the older model thats why you are getting wrong labels let me fix that meanwhile you can try inference using web app

**Comment by ncightadmin** [comment]
also found an issue with web app,It asks to select a surgeo, but there is no surgeon to select and unable to type in the surgeon IDAlonzo Sexton, MDCEO, Founder
        
        
          On Dec 18 2025, at 8:13 pm, muhammadanas25 ***@***.***> wrote:
          
        
        
           muhammadanas25 left a comment (ncightadmin/video-analyzer#1)
Can you specify on which video file this issue was reported so I could reproduce it and that would help me better to understand and solve this issue?
Also can you use the web app to do inference ? its is the same web app  which I shown in demo
the desktop app is currently using the older model thats why you are getting wrong labels let me fix that meanwhile you can try inference using web app
